packages/core/src/__tests__/state-manager.test.ts

```diff
@@ -71,75 +89,151 @@ describe('anchor', () => {
     ;(stateManager as any)._handleTip = fakeHandleTip
+
+    // Mock a bunch of throws as consecutive calls to fakeHandleTip
     fakeHandleTip.mockRejectedValueOnce(new Error('Handle tip failed'))
     fakeHandleTip.mockRejectedValueOnce(new Error('Handle tip failed'))
     fakeHandleTip.mockRejectedValueOnce(new Error('Handle tip failed'))
-    fakeHandleTip.mockImplementationOnce(realHandleTip)
+    fakeHandleTip.mockRejectedValueOnce(new Error('Handle tip failed'))
+    fakeHandleTip.mockRejectedValueOnce(new Error('Handle tip failed'))
+    fakeHandleTip.mockRejectedValueOnce(new Error('Handle tip failed'))
+
+    // wait for 3 secs to let the retry mechanism work
+    await TestUtils.delay(3000)
+
+    // Check that fakeHandleTip was called only three times
+    expect(fakeHandleTip).toHaveBeenCalledTimes(3)
+  })
+
+  test('anchor request is stored when created and deleted when anchored', async () => {
+    const stream = await TileDocument.create(ceramic, INITIAL_CONTENT, null, { anchor: false })
+    const stream$ = await ceramic.repository.load(stream.id, {})
+
+    // @ts-ignore anchorRequestStore is private
+    const anchorRequestStore = ceramic.repository.stateManager.anchorRequestStore
+
+    expect(await anchorRequestStore.load(stream.id)).toBeNull()
+    await ceramic.repository.stateManager.anchor(stream$)
+    expect(stream$.value.anchorStatus).toEqual(AnchorStatus.PENDING)
+    expect(await anchorRequestStore.load(stream.id)).not.toBeNull()
+    await waitForAnchorStatus(stream$, AnchorStatus.ANCHORED)
+    expect(stream$.value.anchorStatus).toEqual(AnchorStatus.ANCHORED)
+    await TestUtils.delay(100) // Needs a bit of time delete the request from the store
+    expect(await anchorRequestStore.load(stream.id)).toBeNull()
+  })
 
-    await new Promise<void>((resolve) => {
-      ceramic.repository.stateManager.anchor(stream$).add(resolve)
+  test('anchor request is stored when created and deleted when failed', async () => {
+    // create stream without requesting anchor
+    const ceramicWithManualAnchoring = await createCeramic(ipfs, { anchorOnRequest: false })
+    const stream = await TileDocument.create(ceramicWithManualAnchoring, { x: 1 }, null, {
+      anchor: true,
     })
-    expect(stream$.value.anchorStatus).toEqual(AnchorStatus.FAILED)
+    expect(stream.state.anchorStatus).toEqual(AnchorStatus.PENDING)
+
+    // @ts-ignore anchorRequestStore is private
+    const anchorRequestStore = ceramicWithManualAnchoring.repository.stateManager.anchorRequestStore
+
+    expect(await anchorRequestStore.load(stream.id)).not.toBeNull()
+
+    // fail anchor
+    const anchorService = ceramicWithManualAnchoring.context
+      .anchorService as any as InMemoryAnchorService
+    await anchorService.failPendingAnchors()
+    await stream.sync()
+    expect(stream.state.anchorStatus).toEqual(AnchorStatus.FAILED)
+    await TestUtils.delay(100) // Needs a bit of time delete the request from the store
```

Comment:
I hate how racy this is.  It might be time for us to add a `TestUtils.assertSoon` method that takes a function and calls it periodically until some timeout, and errors if the function doesn't return true within the timeout.

Reply:
Okay I went ahead and filed CDB-2090.  This can go in with the sleep approach, but let's add a TODO referencing that ticket and I'll clean it up with the helper once I finish it.

Reply:
PR here: https://github.com/ceramicnetwork/js-ceramic/pull/2585.  You should be able to use the new `waitForConditionOrTimeout` function added there to remove these calls to `TestUtils.delay`